### DISCUSSION TITLE
How do you update a user?

### DISCUSSION THREAD

**Discussion by adamzerner** [supabase/supabase#11693] [Questions]
I am following the [Next.js guide](https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs). It [says](https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs#set-up-the-database-schema) to use the User Management Starter. That starter has a Users table in the auth schema with a column of `raw_user_meta_data` of type `jsonb`. It also has a Profiles table in the public schema that kinda corresponds to this `raw_user_meta_data` stuff. When you create a user it has a trigger that creates a row in the Profiles table for you automatically.

But what about updating a user? If you're updating the email or password [updateUser](https://supabase.com/docs/reference/javascript/auth-updateuser) would work, but if you're updating one of the metadata fields and use `updateUser` with `data`, it'll only update `raw_user_meta_data` in Users, not the corresponding row in Profiles. I'd think you'd want the row in Profiles to update as well, otherwise what's the point of having it? I'd write my own trigger to do this but I don't understand how.

Alternatively you could update Profiles directly doing something like `supabase.from('profiles').upsert(updates)`. That's what the Next.js tutorial [does](https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs#account-page). But then the metadata in the Users table gets out of sync with Profiles which seems unwise. Also it means that `useSession` and `useUser` give you the wrong info, which is inconvenient.

I guess you could and probably should be doing both: 1) `updateUser` and 2) `supabase.from('profiles').upsert(updates)`.

I'm interested to hear what would be an advisible way of handling user updates.

Note: I read through https://github.com/supabase/supabase-js/issues/23 but it doesn't really address what I'm looking for: a discussion of these tradeoffs and a solution that is best.

Update: Actually, I'm wondering whether it'd make sense to just avoid the `public.profiles` table all together and use `auth.users` as the source of truth. Is there a downside to that? The only one I see is that `auth.users` has the domain-specific stuff in the `raw_user_meta_data` column, which is `jsonb`, which can make certain queries more difficult. But I think that's probably fine for a lot of use cases. The upside I see is less "surface area" for things to go wrong (if you need to keep `public.profiles` and `auth.users` in sync things can go wrong, ie when you add columns and need to update the trigger).

**Comment by GaryAustin1** [comment]
You can add an update trigger and function (you can actually have a function handle both) to auth.users and update any data that changes in the users row to your profile table.

This has examples of using "TG_OP" to have a single function handle multiple types of triggers.  https://www.postgresql.org/docs/current/plpgsql-trigger.html

**Reply by adamzerner** [reply]
Yeah, that does seem like the best solution. You want `auth.users` and `public.profiles` to remain in sync and a trigger is probably the most reliable way of doing that. And it's consistent with how there is currently a trigger that does this when a record in `auth.users` is created; it seems a little weird for there to be a trigger for that but a different way to keep things in sync after updates.

I guess my problem then is that this PL/pgSQL syntax is not something that I'm familiar with and don't really feel comfortable writing code in. I suspect that the same is true for a lot of other people following the [advice](https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs#set-up-the-database-schema) of the Next.js guide to use the User Management Starter, in which case it'd be nice if there was a sort of plug-n-play solution. Would you be able to advise?

Currently for the User Management Starter it provides a `handle_new_user` function for the `public` schema:
```
begin
  insert into public.profiles (id, first_name, last_name, avatar_url)
  values (new.id, new.raw_user_meta_data->>'first_name', new.raw_user_meta_data->>'last_name', new.raw_user_meta_data->>'avatar_url');
  return new;
end;
```
And an `on_auth_user_created` trigger for the `auth` schema and `users` table that calls that `handle_new_user` function on the `AFTER INSERT` event.

**Reply by adamzerner** [reply]
Note: I wrote an update to the OP questioning whether it makes sense to have two sources of truth (`auth.users` and `public.profiles`) in the first place.

**Reply by peppermintbunny** [reply]
Can I ask - are you saying there is a trigger so that when a user is added to auth.users are record is also added for that user in public.profiles, presumably with the user's uid?  This would solve a problem I have!

**Reply by GaryAustin1** [reply]
> Can I ask - are you saying there is a trigger so that when a user is added to auth.users are record is also added for that user in public.profiles, presumably with the user's uid? This would solve a problem I have!

There is not a trigger/function but you can write one.

**Reply by GaryAustin1** [reply]
https://supabase.com/docs/guides/auth/managing-user-data#advanced-techniques

**Reply by adamzerner** [reply]
Actually when you use the [User Management Starter](https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs#set-up-the-database-schema) it creates one for you. But only for create, not update or delete.

**Reply by peppermintbunny** [reply]
The problem with the option in the User Management Starter is that it uses a method to create a new user AND once the user is created, that user is logged in.  I'm looking for a way for user1 to create a new user, user2, and still remain as the logged in user (i.e. user1).  This is for a tenancy type app.

**Comment by peppermintbunny** [comment]
Gary, 
Thanks for answering.  Your links were very useful and helped me find a solution to my problem. 🤗
